### PR TITLE
A quick implementation of activation interceptors.

### PR DESCRIPTION
**Pull request by AaronAllBright**
A quick implementation of activation interceptors. Would love some feedback.

Possibly addressing issue #9

### CODE REVIEW

**Comment by jeremydmiller** [comment]
Oh man, I'm so sorry about how long it's taken to respond.

The approach looks fine to me, but I'd want to see some tests to understand the usability next

**Comment by AaronAllBright** [comment]
I believe I did include a couple tests.

**Comment by paziz001** [comment]
I copied this files in my solution and the interceptor works fine but for some reason the lifetime of the intercepting instances doesn't follow the lifetime of the inner instances.

**Comment by paziz001** [comment]
I found the place where that happens
https://github.com/JasperFx/lamar/blob/master/src/Lamar/ServiceFamily.cs#L76

**Comment by Cybrosys** [comment]
This helped me implement a DynamicProxy trace logging solution, this should definitely be merged 👍

File: src/Lamar/IoC/Instances/InterceptingInstance.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+        public override void Reset()
+        {
+            base.Reset();
+            _inner.Reset();
+        }
+
+        public override Variable CreateVariable(BuildMode mode, ResolverVariables variables, bool isRoot) => new GetInstanceFrame(this).Variable;
+
+        public override Func<Scope, object> ToResolver(Scope topScope) => s => Resolve(topScope);
```
Comment (Cybrosys) [inline src/Lamar/IoC/Instances/InterceptingInstance.cs:40]: `Resolve(topScope)` should be `Resolve(s)`.

Comment (Cybrosys) [reply src/Lamar/IoC/Instances/InterceptingInstance.cs:40]: @jeremydmiller Just got the notification that you merged this, please see above. We had an issue which was resolved by changing:
`public override Func<Scope, object> ToResolver(Scope topScope) => s => Resolve(topScope);`
to:
`public override Func<Scope, object> ToResolver(Scope topScope) => s => Resolve(s);`

I'd love to give some more details but my memory is a bit hazy at the moment.

**Pull request merged**